rfcs/0006-configuration-constants.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: Configuration Time Constants, `config_time_constants`
+- Start Date: 2022-10-01
+- RFC PR: [FuelLabs/sway-rfcs#0006](https://github.com/FuelLabs/sway-rfcs/pull/19)
+- Sway Issue: [FueLabs/sway#1498](https://github.com/FuelLabs/sway/issues/1498)
+
+# Summary
+
+[summary]: #summary
+
+Configuration time constants can be conceptualized as traditional environment variables. Some bytecode has been compiled by the Sway compiler, and the SDK would like to configure some behavior of that bytecode with additional inputs that won't trigger a recompile.
+
+There is a similar feature that was implemented [here](https://github.com/FuelLabs/sway/pull/2549), but that was a mistaken interpretation of the requirements. In #2549, a recompile still occurs and the new values are injected via `Forc.toml`, when we actually want these values to be injectable by the SDK.
+
+# Motivation
+
+[motivation]: #motivation
+
+This supports the use case of configurable values post-compile-time but pre-runtime, similar to Solidity's `immutable` concept or environment variables in traditional programming. This allows for, e.g., updating contract addresses without recompiling the bytecode.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+When a Sway program is compiled, a set of artifacts are produced. Included in these artifacts are a JSON ABI descriptor, a configuration-time-constant descriptor, and the actual bytecode. The object we are discussing here is the configuration-time-constant descriptor, which describes the specific offset within the bytecode, in terms of bytes, to the data section entry for a particular configuration variable.
+
+The configuration variables for a program can be identified by the `configurable` keyword within the Sway language. This denotes a constant within the program that has no known value at compile time. The value will be provided at configuration time by the SDK or some other consumer of the configuration-time-constant descriptor. As an example:
```

Comment:
I personally do like `configurable`, because it reads like the value is a configurable variable. `config` sounds stateful to me, like a config file or something. Happy to bikeshed though